cleverhans/attacks.py

```diff
@@ -48,48 +65,98 @@ def generate(self, x, **kwargs):
         if self.back == 'th':
             raise NotImplementedError('Theano version not implemented.')
 
-        if not self.inf_loop:
-            self.inf_loop = True
-            assert self.parse_params(**kwargs)
-            import tensorflow as tf
-            graph = tf.py_func(self.generate_np, [x], tf.float32)
-            self.inf_loop = False
-            return graph
-        else:
-            error = "No symbolic or numeric implementation of attack."
-            raise NotImplementedError(error)
+        error = "Sub-classes must implement generate."
+        raise NotImplementedError(error)
 
     def generate_np(self, x_val, **kwargs):
         """
-        Generate adversarial examples and return them as a Numpy array. This
-        method should be overriden in any child class that implements an attack
-        that is not fully expressed symbolically.
+        Generate adversarial examples and return them as a Numpy array.
+        Sub-classes *should not* implement this method unless they must
+        perform special handling of arguments.
         :param x_val: A Numpy array with the original inputs.
         :param **kwargs: optional parameters used by child classes.
         :return: A Numpy array holding the adversarial examples.
         """
         if self.back == 'th':
             raise NotImplementedError('Theano version not implemented.')
 
-        if not self.inf_loop:
-            self.inf_loop = True
-            import tensorflow as tf
-
-            # Generate this attack's graph if not done previously
-            if not hasattr(self, "_x") and not hasattr(self, "_x_adv"):
-                input_shape = list(x_val.shape)
-                input_shape[0] = None
-                self._x = tf.placeholder(tf.float32, shape=input_shape)
-                self._x_adv = self.generate(self._x, **kwargs)
-            self.inf_loop = False
-        else:
-            error = "No symbolic or numeric implementation of attack."
-            raise NotImplementedError(error)
+        import tensorflow as tf
 
         if self.sess is None:
             raise ValueError("Cannot use `generate_np` when no `sess` was"
                              " provided")
-        return self.sess.run(self._x_adv, feed_dict={self._x: x_val})
+
+        # the set of arguments that are structural properties of the attack
+        # if these arguments are different, we must construct a new graph
+        fixed = dict((k, v) for k, v in kwargs.items()
```

Comment:
I set

    hash_key = tuple(sorted(fixed.items()))

so that it should be identical between runs. But if you think ordereddict is better, we could do that.